sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/aggregator/DoubleHistogramAggregator.java

```diff
@@ -0,0 +1,158 @@
+    super(resource, instrumentationLibraryInfo, instrumentDescriptor, stateful);
+    this.boundaries = boundaries;
+
+    List<Double> boundaryList = new ArrayList<>(this.boundaries.length);
+    for (double v : this.boundaries) {
+      boundaryList.add(v);
+    }
+    this.boundaryList = Collections.unmodifiableList(boundaryList);
+  }
+
+  @Override
+  public AggregatorHandle<HistogramAccumulation> createHandle() {
+    return new Handle(this.boundaries);
+  }
+
+  /**
+   * Return the result of the merge of two histogram accumulations. As long as one Aggregator
+   * instance produces all Accumulations with constant boundaries we don't need to worry about
+   * merging accumulations with different boundaries.
+   */
+  @Override
+  public final HistogramAccumulation merge(HistogramAccumulation x, HistogramAccumulation y) {
+    long[] mergedCounts = new long[x.getCounts().length];
+    for (int i = 0; i < x.getCounts().length; ++i) {
+      mergedCounts[i] = x.getCounts()[i] + y.getCounts()[i];
+    }
+    return HistogramAccumulation.create(x.getSum() + y.getSum(), mergedCounts);
+  }
+
+  @Override
+  public final MetricData toMetricData(
+      Map<Labels, HistogramAccumulation> accumulationByLabels,
+      long startEpochNanos,
+      long lastCollectionEpoch,
+      long epochNanos) {
+    return MetricData.createDoubleHistogram(
+        getResource(),
+        getInstrumentationLibraryInfo(),
+        getInstrumentDescriptor().getName(),
+        getInstrumentDescriptor().getDescription(),
+        getInstrumentDescriptor().getUnit(),
+        DoubleHistogramData.create(
+            isStateful() ? AggregationTemporality.CUMULATIVE : AggregationTemporality.DELTA,
+            MetricDataUtils.toDoubleHistogramPointList(
+                accumulationByLabels,
+                isStateful() ? startEpochNanos : lastCollectionEpoch,
+                epochNanos,
+                boundaryList)));
+  }
+
+  @Override
+  public HistogramAccumulation accumulateDouble(double value) {
+    long[] counts = new long[this.boundaries.length + 1];
+    counts[findBucketIndex(this.boundaries, value)] = 1;
+    return HistogramAccumulation.create(value, counts);
+  }
+
+  @Override
+  public HistogramAccumulation accumulateLong(long value) {
+    return accumulateDouble((double) value);
+  }
+
+  // Benchmark shows that linear search performs better than binary search with ordinary
+  // buckets.
+  private static int findBucketIndex(double[] boundaries, double value) {
+    for (int i = 0; i < boundaries.length; ++i) {
+      if (Double.compare(value, boundaries[i]) <= 0) {
+        return i;
+      }
+    }
+    return boundaries.length;
+  }
+
+  static final class Handle extends AggregatorHandle<HistogramAccumulation> {
+    // read-only
+    private final double[] boundaries;
+
+    @GuardedBy("lock")
+    private double sum;
+
+    @GuardedBy("lock")
+    private final long[] counts;
+
+    private final ReentrantLock lock = new ReentrantLock();
```

Comment:
```suggestion
    private final Object mutex = new Object();
```

Reply:
Can you update below to use `synchronized(mutex)`? It's the pattern we use everywhere else so we should be consistent, and I think it might be a bit faster. We'd only use `ReentrantLock` if we used any of the methods other than `lock` / `unlock`

Reply:
To clarify, by "everywhere else" I generally refer to non-metrics since I haven't looked at the metrics code so much. But we should get it closer to our tracing code in these aspects

Reply:
mutex benchmark:
|Benchmark                                          |Mode  |Cnt     |Score     |Error  |Units|
|---                                                       |---      |---       |---          |---     |---|
|DoubleHistogramBenchmark.aggregate_10Threads       |avgt   |10  |1210.696 |±  50.972  |ns/op|
|DoubleHistogramBenchmark.aggregate_1Threads       | avgt  | 10     |9.997 |±   0.448  |ns/op|
|DoubleHistogramBenchmark.aggregate_5Threads        |avgt  | 10  | 696.606| ± 123.858  |ns/op|

lock benchmark:
|Benchmark                                          |Mode  |Cnt     |Score     |Error  |Units|
|---                                                       |---      |---       |---          |---     |---|
|DoubleHistogramBenchmark.aggregate_10Threads       |avgt   |10  |  371.369 |± 8.39 | ns/op|
|DoubleHistogramBenchmark.aggregate_1Threads       |avgt   |10  |   24.045 |± 0.648|  ns/op|
|DoubleHistogramBenchmark.aggregate_5Threads       |avgt   |10  | 187.410| ± 0.863 | ns/op|

the `mutext` approach works better only in a single-threaded environment. not sure if we should use `mutex` here.

Reply:
Hmm - so many articles say they should have at least the same performance but obviously there are some traps too. `ReentrantLock` seems to make sense here